dso-l2/src/main/java/com/tc/objectserver/impl/PeriodicCallable.java

```diff
@@ -112,15 +116,22 @@ public boolean evictionStarted(ObjectID oid) {
   }
 
   @Override
-  public synchronized boolean evictionCompleted(ObjectID oid) {
-    listeningSet.remove(oid);
-    if ( listeningSet.isEmpty() ) {
-      if ( !stopped && current == null && !workingSet.isEmpty() ) {
-          evictor.schedulePeriodicEvictionRun(workingSet);
+  public boolean evictionCompleted(ObjectID oid) {
+    boolean complete, reschedule = false;
+    synchronized (this) {
+      listeningSet.remove(oid);
+      if (listeningSet.isEmpty()) {
+        if (!stopped && current == null && !workingSet.isEmpty()) {
+          reschedule = true;
+        }
+        complete = true;
+      } else {
+        complete = false;
       }
-      return true;
-    } else {
-      return false;
     }
+    if (reschedule) {
+      evictor.schedulePeriodicEvictionRun(workingSet);
```

Comment:
There is a potential bug here, in that we're pulling the workingSet out of it's guarded lock scope and handing it off to another object. We've always been doing this however, so this shouldn't introduce any new buggy behavior. If someone wants to argue for fixing it then I can do so by making a defensive copy under lock.

Reply:
is the effect of the bug simply that something that would have been evicted isn't?

Reply:
Well... absolute worst case would be a corruption of the internal state of the `BitSetObjectIDSet`. My strong suspicion though is that there is no concurrent access and that this is a clean handoff. Proving it might be tricky though.

Reply:
I don't have all of the code checked out (at the moment) but there is a copy constructor for `BitSetObjectIDSet`  so I'd use it.  It makes sense to me to copy `workingSet` within the sync block and then pass it along when rescheduling eviction.  I'm not sure its safe to extend the life of the current `workingSet` object outside the scope of the current `PeriodicCallable`.

Reply:
Other than that, the changes look like they should address the deadlock.